File: 0206-reverse-linked-list/0206-reverse-linked-list.c
// Aug 11, 2026 10:51
// three pointer approach

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* reverseList(struct ListNode* head) {
    // edge case
    if (head == NULL) {
        return NULL;
    }

    // we set three pointers
    struct ListNode *prev = NULL;
    struct ListNode *curr = head;
    struct ListNode * next = NULL;

    // now we reverse the linked list in place
    while (curr != NULL) {
        next = curr->next;
        curr->next = prev;
        prev = curr;
        curr = next;
    }

    head = prev;

    return head;
}

File: 0206-reverse-linked-list/0206-reverse-linked-list.py
# Aug 11, 2026 10:47
# three pointer approach

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # edge case
        if head is None:
            return None
        
        # we set three pointers
        # prev will point to the previous node, initially null
        # curr will initially point to the head
        # next will point to the next node so we can keep track of the linked list
        prev = None
        curr = head
        next = None

        # now we modify the connections of the nodes in place
        while curr is not None:
            next = curr.next
            curr.next = prev
            prev = curr
            curr = next
        
        head = prev
        
        return head
